core/state/statedb.go

```diff
@@ -472,12 +530,15 @@ func (s *StateDB) commit(dbw trie.DatabaseWriter) (root common.Hash, err error)
 			}
 			// Update the object in the main account trie.
 			s.UpdateStateObject(stateObject)
-			s.all[addr] = stateObject.data
 		}
 		delete(s.stateObjectsDirty, addr)
 	}
 	// Write trie changes.
-	return s.trie.CommitTo(dbw)
+	root, err = s.trie.CommitTo(dbw)
+	if err == nil {
+		s.pushTrie(s.trie)
```

Comment:
So, `pastTries` will contain the most recent n tries with changes; not necessarily the n most recent blocks' tries. Is this intentional? Could the attacker start submitting, say, forks of ancient blocks to force us to replace the trie cache?